### ISSUE TITLE
Constraints on Vectors

### ISSUE THREAD

**Issue by smukoehler** [JuliaOpt/JuMP.jl#403]
Upon browsing the documentation at http://jump.readthedocs.org/en/latest/refexpr.html#constraint-references for how to write constraints in JuMP, I couldn't find an easy way to write constraints that are placed on vectors rather than scalars. For example, I tried doing this in JuMP:

```
A = rand(2,2); 
B = rand(2,2); # not usually random, just used for illustrative purposes
x0 = [1;1];
controlProblem = Model();
@defVar(controlProblem,x[1:2])
@defVar(controlProblem,u[1:2])
@addConstraint(controlProblem, x==A*x0+B*u)
```
 and got the following error:
```
Cannot construct an affine expression with a term of type (Float64)*(JuMPArray##8569{Variable})
while loading In[76], in expression starting on line 319

 in error at error.jl:21
 in addToExpression at /Users/skoehler/.julia/v0.3/JuMP/src/macros.jl:97
```

I also looked through the tutorials for JuMP and couldn't find anything like this. Am I just getting the syntax wrong? I could split this up into a for loop, but it is not a very pretty solution. Maybe @tkelman knows something about this?

**Comment by mlubin** [comment]
We don't currently have matrix syntax like this (though Convex.jl does). Here's a one-liner that does the same thing:
```
@addConstraint(controlProblem, x_eq[i=1:size(A,1)], x[i] == dot(A[i,:],x0) + dot(B[i,:],u))
```

**Comment by tkelman** [comment]
Hey Sarah. Hrm, that is a liitle ugly and annoying to have to split everything out as scalars. It looks like `JuMPArray`s don't support `size` and since indexing into them can have an offset and start at something other than 1, the generic array operators in base Julia won't necessarily work for them...

**Comment by IainNZ** [comment]
We're not against the idea necessarily, but given how general variable indexing can be in JuMP it definitely takes some care. The other thing is that we've found fairly often that people don't have dense block matrices to add to a problem, so "spelling it out" is often the way to go anyway.

**Comment by tkelman** [comment]
In control problems we very very frequently have block matrices, at least for linear problems. They may not always be dense block matrices, but if `A` or `B` were sparse in the above example then Miles' workaround would give
```
ERROR: `dot` has no method matching dot(::SparseMatrixCSC{Float64,Int64}, ::Array{Float64,1})
 in anonymous at /home/tkelman/.julia/v0.3/JuMP/src/macros.jl:143
```
(yes it would probably be better to traverse the CSC data structure directly, but not everyone is going to know how to do that)

**Comment by mlubin** [comment]
We can implement Matrix{Float64}*JuMPArray{Variable} and related overloads and just throw an error when the JuMPArray doesn't use simple 1-based indices.

**Comment by IainNZ** [comment]
I guess @joehuchette has walked that path to a degree on the SDP branch - is there any pitfall we aren't thinking of? I know we've punted on it before to a degree but now might be the time.

**Comment by joehuchette** [comment]
I punted a little bit by introducing new ``MatrixVar`` types that I could use dispatch to treat differently. We can definitely add overloads for ``A*x`` that verifies indexing is kosher, but I feel like that is a stopgap that would be better handled by the type system. Maybe a boolean type parameter that is true for JuMPArrays with indexing like ``x[1:n1,1:n2,1:n3]``?

**Comment by mlubin** [comment]
@joehuchette, a boolean type parameter makes sense, I think we could do that.

**Comment by joehuchette** [comment]
I also think a boolean type parameter for square matrices would be nice for SDPs, but that's more speculative.

**Comment by smukoehler** [comment]
Thanks guys. I'm glad to see this was a worthwhile issue to point out. The example I wrote is something you would see in any optimal control problem. I look forward to seeing your solution.

**Comment by joehuchette** [comment]
Thanks for opening, we've talked about this forever but sometimes it takes a kick to actually get started.

**Comment by mlubin** [comment]
@smukoehler, in the meantime you may want to try out Convex.jl if you want to write your whole problem in vectorized form.

**Comment by mlubin** [comment]
Closed by #404. Use ``Pkg.checkout("JuMP")`` to test it out.
